src/transformStateWithClones.js

```diff
@@ -6,8 +6,27 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = [];
+  let stateCopy = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      stateCopy = {};
+    } else if (action.type === 'addProperties') {
+      stateCopy = { ...stateCopy, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      stateCopy = { ...stateCopy };
+
+      action.keysToRemove.forEach((key) => {
+        delete stateCopy[key];
+      });
+    }
+    newState.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF] - It is better to push a copy of the object to the array at the end of the each loop cycle, but outside of the `switch` block, after current action is processed. This will avoid duplication of the object spread operation `{ ...stateCopy }` for each case.